gordon/record_checker.py

```diff
@@ -0,0 +1,112 @@
+import asyncio
+import logging
+import time
+
+from async_dns import types
+from async_dns.resolver import ProxyResolver
+
+
+class RecordChecker(object):
+    QUERYING_THRESHOLD = 60
+
+    def __init__(self, dns_ip):
+        """Setup RecordChecker object
+
+        Args:
+            dns_ip: DNS ip server to query
+        """
+        self._dns_ip = dns_ip
+        self._resolver = ProxyResolver()
+        self._resolver.set_proxies([self._dns_ip])
+
+    def _extract_record_data(self, record):
+        """Extract record data from a record dict
+
+        Args:
+            record (dict): DNS record as a dict
+        """
+        return record.values()
+
+    async def check_record(self, record):
+        """Asynchronously queries a DNS server for a given record
+            to check how long it takes the record to become available.
+
+        Args:
+            record (dict): DNS record as a dict
+            with record properties
+
+        """
+        start_time = time.time()
+
+        name, rr_data, r_type, ttl = self._extract_record_data(record)
+        r_type_code = types.get_code(r_type)
+
+        record_found_in_dns_srv = False
+        query_tries = 0
+
+        while not record_found_in_dns_srv and\
```

Comment:
What do you think about making this function wait exponentially before each try (exponential back off)? I'm thinking about the scenario when there are suddenly 30 records to check, and each of these records makes a DNS query every second. Do you think this is a legit concern or premature optimization?

Reply:
@matt0bi wouldn't introducing an exponential backoff misrepresent the actual resolution time of the record? i could see a case where a large exponential backoff is in place, but the record was resolvable well before we re-attempt. i'd hope the upstream resolvers could handle these metric (once a second) queries fine. either way, do we have an idea of how granular we would want to look at this data or what SLO it would inform?

Reply:
This is a good point @smstone. I'm sorry @nuritizra for leading you down a wrong path, I think my suggestion to use exponential back off wasnt right . I talked with @smstone about this over slack we came up with an idea that the check should retry every 5 seconds, which should be enough granularity. But we're not 100% sure what the timeout value should be, because this is tied to how often the publisher plugin will re-POST the request. Maybe we can talk about this tomorrow?